app/celery/process_pinpoint_receipts_tasks.py

```diff
@@ -0,0 +1,145 @@
+            return
+
+        if notification.status != NOTIFICATION_SENT:
+            notifications_dao._duplicate_update_warning(notification, notification_status)
+            return
+
+        notifications_dao._update_notification_status(
+            notification=notification,
+            status=notification_status,
+            provider_response=provider_response,
+        )
+
+        if notification_status != NOTIFICATION_DELIVERED:
+            current_app.logger.info(
+                (
+                    f"Pinpoint delivery failed: notification id {notification.id} and reference {reference} has error found. "
+                    f"Provider response: {provider_response}"
+                )
+            )
+        else:
+            current_app.logger.info(
+                f"Pinpoint callback return status of {notification_status} for notification: {notification.id}"
+            )
+
+        statsd_client.incr(f"callback.sns.{notification_status}")  # TODO: do we want a Pinpoint metric here?
+
+        if notification.sent_at:
+            statsd_client.timing_with_dates(
+                "callback.sns.elapsed-time", datetime.utcnow(), notification.sent_at
+            )  # TODO: do we want a Pinpoint metric here?
+
+        _check_and_queue_callback_task(notification)
+
+        return True
+
+    except Retry:
+        raise
+
+    except Exception as e:
+        current_app.logger.exception(f"Error processing Pinpoint results: {str(e)}")
+        self.retry(queue=QueueNames.RETRY)
+
+
+def determine_pinpoint_status(sns_status, provider_response):
+    if sns_status == "DELIVERED":
+        return NOTIFICATION_DELIVERED
+
+    # See all the possible provider responses
+    # https://docs.aws.amazon.com/sns/latest/dg/sms_stats_cloudwatch.html#sms_stats_delivery_fail_reasons
+    reasons = {
+        "Blocked as spam by phone carrier": NOTIFICATION_TECHNICAL_FAILURE,
+        "Destination is on a blocked list": NOTIFICATION_TECHNICAL_FAILURE,
+        "Invalid phone number": NOTIFICATION_TECHNICAL_FAILURE,
+        "Message body is invalid": NOTIFICATION_TECHNICAL_FAILURE,
+        "Phone carrier has blocked this message": NOTIFICATION_TECHNICAL_FAILURE,
+        "Phone carrier is currently unreachable/unavailable": NOTIFICATION_TEMPORARY_FAILURE,
+        "Phone has blocked SMS": NOTIFICATION_TECHNICAL_FAILURE,
+        "Phone is on a blocked list": NOTIFICATION_TECHNICAL_FAILURE,
+        "Phone is currently unreachable/unavailable": NOTIFICATION_PERMANENT_FAILURE,
+        "Phone number is opted out": NOTIFICATION_PERMANENT_FAILURE,
+        "This delivery would exceed max price": NOTIFICATION_TECHNICAL_FAILURE,
+        "Unknown error attempting to reach phone": NOTIFICATION_TECHNICAL_FAILURE,
+    }
+
+    status = reasons.get(provider_response)  # could be None
+    if not status:
+        # TODO: Pattern matching in Python 3.10 should simplify this overall function logic.
```

Comment:
This was copied over older code I assume? we can try this pattern matching now? 👀

Reply:
I did some matching but I think it's cleaner refactored into an if/elif block. 🤔